### ISSUE TITLE
Build Error on Ubuntu 18.04

### ISSUE THREAD

**Issue by wissn84** [project-chip/connectedhomeip#5127]
#### Problem
- Build Error on Ubuntu 18.04
- I think this  commit( [5ca7308145faf5100592627024c30c93d0d7498a))](https://github.com/project-chip/connectedhomeip/commit/5ca7308145faf5100592627024c30c93d0d7498a) cause of Build Error 
- When I roll back before the commit, I have checked build successful.

Build Error Log is below

 #### Proposed Solution
`ninja -C out/debug
ninja: Entering directory `out/debug'
[21/73] c++ obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.MainLoop.cpp.o
FAILED: obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.MainLoop.cpp.o
g++ -MMD -MF obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.MainLoop.cpp.o.d -Wconversion -O0 -g2 -fno-common -ffunction-sections -fdata-sections -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fPIC -Wall -Wextra -Wshadow -Wunreachable-code -Werror -Wno-deprecated-declarations -Wno-unknown-warning-option -Wno-missing-field-initializers -Wno-unused-parameter -Wno-psabi -Wno-cast-function-type -fdiagnostics-color -fno-strict-aliasing -D_REENTRANT -isystem../../third_party/connectedhomeip/third_party/ot-br-posix/repo/src -isystem../../third_party/connectedhomeip/third_party/ot-br-posix/repo/include -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gio-unix-2.0/ -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -std=gnu++14 -fno-rtti -Wnon-virtual-dtor -DCHIP_HAVE_CONFIG_H=1 -DCHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE=true -DLWIP_IPV4=1 -DLWIP_IPV6=1 -DLWIP_API=1 -DLWIP_ETHERNET=1 -DLWIP_SLIP=0 -DLWIP_6LOWPAN=1 -DLWIP_PPP=1 -I../../third_party/connectedhomeip/src/include -I../../third_party/connectedhomeip/src -Igen/include -I../../third_party/connectedhomeip/config/standalone -I../../third_party/connectedhomeip/src/lib -I../../third_party/connectedhomeip/third_party/nlassert/repo/include -I../../third_party/connectedhomeip/third_party/nlio/repo/include -I../../third_party/connectedhomeip/src/lwip/standalone -I../../third_party/connectedhomeip/third_party/lwip/repo/lwip/src/include -I../../third_party/connectedhomeip/src/lwip/include -I../../third_party/connectedhomeip/src/setup_payload -Igen -Igen -I../../third_party/connectedhomeip/third_party/inipp/repo/inipp -c ../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.cpp -o obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.MainLoop.cpp.o
In file included from ../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.cpp:17:0:
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h: In instantiation of ‘bool chip::DeviceLayer::Internal::MainLoop::Schedule(int (*)(T*), T*) [with T = chip::DeviceLayer::Internal::{anonymous}::CallbackIndirection]’:
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.cpp:171:63:   required from here
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: error: ‘G_SOURCE_FUNC’ was not declared in this scope
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: note: suggested alternative: ‘G_STRFUNC’
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
                                 G_STRFUNC
cc1plus: error: unrecognized command line option ‘-Wno-cast-function-type’ [-Werror]
cc1plus: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]
cc1plus: all warnings being treated as errors
[23/73] c++ obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.Helper.cpp.o
FAILED: obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.Helper.cpp.o
g++ -MMD -MF obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.Helper.cpp.o.d -Wconversion -O0 -g2 -fno-common -ffunction-sections -fdata-sections -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fPIC -Wall -Wextra -Wshadow -Wunreachable-code -Werror -Wno-deprecated-declarations -Wno-unknown-warning-option -Wno-missing-field-initializers -Wno-unused-parameter -Wno-psabi -Wno-cast-function-type -fdiagnostics-color -fno-strict-aliasing -D_REENTRANT -isystem../../third_party/connectedhomeip/third_party/ot-br-posix/repo/src -isystem../../third_party/connectedhomeip/third_party/ot-br-posix/repo/include -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gio-unix-2.0/ -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -std=gnu++14 -fno-rtti -Wnon-virtual-dtor -DCHIP_HAVE_CONFIG_H=1 -DCHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE=true -DLWIP_IPV4=1 -DLWIP_IPV6=1 -DLWIP_API=1 -DLWIP_ETHERNET=1 -DLWIP_SLIP=0 -DLWIP_6LOWPAN=1 -DLWIP_PPP=1 -I../../third_party/connectedhomeip/src/include -I../../third_party/connectedhomeip/src -Igen/include -I../../third_party/connectedhomeip/config/standalone -I../../third_party/connectedhomeip/src/lib -I../../third_party/connectedhomeip/third_party/nlassert/repo/include -I../../third_party/connectedhomeip/third_party/nlio/repo/include -I../../third_party/connectedhomeip/src/lwip/standalone -I../../third_party/connectedhomeip/third_party/lwip/repo/lwip/src/include -I../../third_party/connectedhomeip/src/lwip/include -I../../third_party/connectedhomeip/src/setup_payload -Igen -Igen -I../../third_party/connectedhomeip/third_party/inipp/repo/inipp -c ../../third_party/connectedhomeip/src/platform/Linux/bluez/Helper.cpp -o obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.Helper.cpp.o
In file included from ../../third_party/connectedhomeip/src/platform/Linux/bluez/Helper.cpp:76:0:
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h: In instantiation of ‘bool chip::DeviceLayer::Internal::MainLoop::SetCleanupFunction(int (*)(T*), T*) [with T = chip::DeviceLayer::Internal::BluezEndpoint]’:
../../third_party/connectedhomeip/src/platform/Linux/bluez/Helper.cpp:1347:79:   required from here
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:83:41: error: ‘G_SOURCE_FUNC’ was not declared in this scope
         mCleanup         = G_SOURCE_FUNC(callback);
                            ~~~~~~~~~~~~~^~~~~~~~~~
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:83:41: note: suggested alternative: ‘G_STRFUNC’
         mCleanup         = G_SOURCE_FUNC(callback);
                            ~~~~~~~~~~~~~^~~~~~~~~~
                            G_STRFUNC
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h: In instantiation of ‘bool chip::DeviceLayer::Internal::MainLoop::Schedule(int (*)(T*), T*) [with T = chip::DeviceLayer::Internal::ConnectionDataBundle]’:
../../third_party/connectedhomeip/src/platform/Linux/bluez/Helper.cpp:1420:101:   required from here
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: error: ‘G_SOURCE_FUNC’ was not declared in this scope
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: note: suggested alternative: ‘G_STRFUNC’
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
                                 G_STRFUNC
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h: In instantiation of ‘bool chip::DeviceLayer::Internal::MainLoop::Schedule(int (*)(T*), T*) [with T = chip::DeviceLayer::Internal::BluezEndpoint]’:
../../third_party/connectedhomeip/src/platform/Linux/bluez/Helper.cpp:1460:65:   required from here
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: error: ‘G_SOURCE_FUNC’ was not declared in this scope
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: note: suggested alternative: ‘G_STRFUNC’
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
                                 G_STRFUNC
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h: In instantiation of ‘bool chip::DeviceLayer::Internal::MainLoop::ScheduleAndWait(int (*)(T*), T*) [with T = chip::DeviceLayer::Internal::BluezEndpoint]’:
../../third_party/connectedhomeip/src/platform/Linux/bluez/Helper.cpp:1554:80:   required from here
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:63:53: error: ‘G_SOURCE_FUNC’ was not declared in this scope
         return RunOnBluezThreadAndWait(G_SOURCE_FUNC(callback), value);
                                        ~~~~~~~~~~~~~^~~~~~~~~~
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:63:53: note: suggested alternative: ‘G_STRFUNC’
         return RunOnBluezThreadAndWait(G_SOURCE_FUNC(callback), value);
                                        ~~~~~~~~~~~~~^~~~~~~~~~
                                        G_STRFUNC
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h: In instantiation of ‘bool chip::DeviceLayer::Internal::MainLoop::Schedule(int (*)(T*), T*) [with T = chip::DeviceLayer::Internal::BluezConnection]’:
../../third_party/connectedhomeip/src/platform/Linux/bluez/Helper.cpp:1672:109:   required from here
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: error: ‘G_SOURCE_FUNC’ was not declared in this scope
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: note: suggested alternative: ‘G_STRFUNC’
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
                                 G_STRFUNC
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h: In instantiation of ‘bool chip::DeviceLayer::Internal::MainLoop::Schedule(int (*)(T*), T*) [with T = _BluezDevice1]’:
../../third_party/connectedhomeip/src/platform/Linux/bluez/Helper.cpp:1742:67:   required from here
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: error: ‘G_SOURCE_FUNC’ was not declared in this scope
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: note: suggested alternative: ‘G_STRFUNC’
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
                                 G_STRFUNC
cc1plus: error: unrecognized command line option ‘-Wno-cast-function-type’ [-Werror]
cc1plus: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]
cc1plus: all warnings being treated as errors
[24/73] c++ obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.ChipDeviceScanner.cpp.o
FAILED: obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.ChipDeviceScanner.cpp.o
g++ -MMD -MF obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.ChipDeviceScanner.cpp.o.d -Wconversion -O0 -g2 -fno-common -ffunction-sections -fdata-sections -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fPIC -Wall -Wextra -Wshadow -Wunreachable-code -Werror -Wno-deprecated-declarations -Wno-unknown-warning-option -Wno-missing-field-initializers -Wno-unused-parameter -Wno-psabi -Wno-cast-function-type -fdiagnostics-color -fno-strict-aliasing -D_REENTRANT -isystem../../third_party/connectedhomeip/third_party/ot-br-posix/repo/src -isystem../../third_party/connectedhomeip/third_party/ot-br-posix/repo/include -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gio-unix-2.0/ -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -std=gnu++14 -fno-rtti -Wnon-virtual-dtor -DCHIP_HAVE_CONFIG_H=1 -DCHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE=true -DLWIP_IPV4=1 -DLWIP_IPV6=1 -DLWIP_API=1 -DLWIP_ETHERNET=1 -DLWIP_SLIP=0 -DLWIP_6LOWPAN=1 -DLWIP_PPP=1 -I../../third_party/connectedhomeip/src/include -I../../third_party/connectedhomeip/src -Igen/include -I../../third_party/connectedhomeip/config/standalone -I../../third_party/connectedhomeip/src/lib -I../../third_party/connectedhomeip/third_party/nlassert/repo/include -I../../third_party/connectedhomeip/third_party/nlio/repo/include -I../../third_party/connectedhomeip/src/lwip/standalone -I../../third_party/connectedhomeip/third_party/lwip/repo/lwip/src/include -I../../third_party/connectedhomeip/src/lwip/include -I../../third_party/connectedhomeip/src/setup_payload -Igen -Igen -I../../third_party/connectedhomeip/third_party/inipp/repo/inipp -c ../../third_party/connectedhomeip/src/platform/Linux/bluez/ChipDeviceScanner.cpp -o obj/third_party/connectedhomeip/src/platform/Linux/bluez/libDeviceLayer.ChipDeviceScanner.cpp.o
In file included from ../../third_party/connectedhomeip/src/platform/Linux/bluez/ChipDeviceScanner.cpp:23:0:
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h: In instantiation of ‘bool chip::DeviceLayer::Internal::MainLoop::Schedule(int (*)(T*), T*) [with T = chip::DeviceLayer::Internal::ChipDeviceScanner]’:
../../third_party/connectedhomeip/src/platform/Linux/bluez/ChipDeviceScanner.cpp:128:63:   required from here
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: error: ‘G_SOURCE_FUNC’ was not declared in this scope
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:56:46: note: suggested alternative: ‘G_STRFUNC’
         return RunOnBluezThread(G_SOURCE_FUNC(callback), value);
                                 ~~~~~~~~~~~~~^~~~~~~~~~
                                 G_STRFUNC
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h: In instantiation of ‘bool chip::DeviceLayer::Internal::MainLoop::ScheduleAndWait(int (*)(T*), T*) [with T = chip::DeviceLayer::Internal::ChipDeviceScanner]’:
../../third_party/connectedhomeip/src/platform/Linux/bluez/ChipDeviceScanner.cpp:171:69:   required from here
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:63:53: error: ‘G_SOURCE_FUNC’ was not declared in this scope
         return RunOnBluezThreadAndWait(G_SOURCE_FUNC(callback), value);
                                        ~~~~~~~~~~~~~^~~~~~~~~~
../../third_party/connectedhomeip/src/platform/Linux/bluez/MainLoop.h:63:53: note: suggested alternative: ‘G_STRFUNC’
         return RunOnBluezThreadAndWait(G_SOURCE_FUNC(callback), value);
                                        ~~~~~~~~~~~~~^~~~~~~~~~
                                        G_STRFUNC
cc1plus: error: unrecognized command line option ‘-Wno-cast-function-type’ [-Werror]
cc1plus: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]
cc1plus: all warnings being treated as errors
[30/73] c++ obj/third_party/connectedhomeip/src/transport/libTransportLayer.NetworkProvisioning.cpp.o
ninja: build stopped: subcommand failed.`

**Comment by wissn84** [comment]
To support 'G_SOURCE_FUNC', Ubuntu glib version must be since version 2.58 but 18.04 has glib version 2.56.4.
To fix this issue, You must be upgrade ubuntu 20.04 upper or upgrade glib but many feedback don't recommend upgrade glib version. because glib is strong coupling with OS.

**Comment by wissn84** [comment]
I have checked build success on Ubuntu 20.04
